File: statusEffects.py
"""Reusable enemy-owned status effects applied by player attacks."""

from dataclasses import dataclass
import random

import gameProfile
import items


@dataclass(frozen=True)
class ControlState:
    movement_multiplier: float = 1.0
    attack_delay: float = 0.0
    stunned: bool = False


def _effects(enemy):
    if not hasattr(enemy, "statusEffects"):
        enemy.statusEffects = {}
        enemy.statusDotBuffer = 0.0
        enemy.statusControlResistance = 0.0
    return enemy.statusEffects


def apply(enemy, kind, duration, potency=0.0, stacks=1):
    effects = _effects(enemy)
    is_boss = hasattr(enemy, "bossName")
    if kind == "stun":
        resistance = min(.8, enemy.statusControlResistance + (.18 if is_boss else .08))
        enemy.statusControlResistance = resistance
        duration *= (1.0 - resistance) * (.45 if is_boss else 1.0)
    current = effects.get(kind, {"remaining": 0.0, "potency": 0.0, "stacks": 0})
    current["remaining"] = max(current["remaining"], duration)
    current["potency"] = max(current["potency"], potency)
    current["stacks"] = min(8 if kind == "bleed" else 3,
                            current["stacks"] + stacks)
    effects[kind] = current
    gameProfile.increment_quest("statuses_applied")


def roll_player_hit(enemy, bullet, equipment, rng=None):
    rng = rng or random
    chances = items.status_chances(equipment)
    # Multi-shot/rapid builds retain status value without multiplying it linearly.
    import characterStats as cS
    coefficient = max(.22, 1.0 / max(1.0, cS.projectileCount) ** .5)
    if getattr(bullet, "currCrit", False):
        chances["bleed"] = chances.get("bleed", 0.0) + .025
    tuning = {
        "poison": (4.5, .007),
        "bleed": (3.2, .006),
        "slow": (2.4, .22),
        "daze": (2.2, .28),
        "stun": (.65, 1.0),
    }
    for kind, chance in chances.items():
        if rng.random() <= min(.65, chance * coefficient):
            duration, potency = tuning[kind]
            apply(enemy, kind, duration, potency)


def damage_multiplier(enemy, bullet=None):
    effects = _effects(enemy)
    multiplier = 1.0
    # Poisoned, dazed targets are "opened" and take a modest synergy bonus.
    if "poison" in effects and "daze" in effects:
        multiplier += .08
    # Bleeding targets reward precision builds without making control mandatory.
    if "bleed" in effects and getattr(bullet, "currCrit", False):
        multiplier += min(.20, effects["bleed"]["stacks"] * .025)
    return multiplier


def update(enemy, seconds):
    effects = _effects(enemy)
    if enemy.statusControlResistance > 0:
        enemy.statusControlResistance = max(0.0, enemy.statusControlResistance - seconds * .035)
    dot_per_second = 0.0
    movement = 1.0
    daze = 0.0
    stunned = False
    expired = []
    for kind, effect in effects.items():
        effect["remaining"] -= seconds
        if effect["remaining"] <= 0:
            expired.append(kind)
            continue
        if kind == "poison":
            dot_per_second += max(2.0, enemy.maxHp * effect["potency"]) * effect["stacks"]
        elif kind == "bleed":
            dot_per_second += max(1.0, enemy.maxHp * effect["potency"]) * effect["stacks"]
        elif kind == "slow":
            movement *= max(.45, 1.0 - effect["potency"])
        elif kind == "daze":
            daze = max(daze, effect["potency"])
        elif kind == "stun":
            stunned = True
    for kind in expired:
        del effects[kind]
    if dot_per_second > 0 and enemy.hp > 0:
        enemy.statusDotBuffer += dot_per_second * seconds
        damage = int(enemy.statusDotBuffer)
        if damage:
            enemy.statusDotBuffer -= damage
            enemy.take_damage(damage)
    return ControlState(movement, daze, stunned)


def summary(enemy):
    return tuple(sorted(_effects(enemy)))

File: RotBoiRemastered/Systems/StatusEffects.cs
using RotBoiRemastered.Entities;

namespace RotBoiRemastered.Systems;

public sealed class StatusEffectState
{
    public double Remaining { get; set; }
    public double Potency { get; set; }
    public int Stacks { get; set; }
}

public sealed record StatusControl(double MovementMultiplier = 1, double AttackDelay = 0, bool Stunned = false);

/// <summary>
/// Player-owned ailments. Periodic damage is explicitly tagged as damage over
/// time, so it can hurt bosses without pretending every tick was a fresh hit.
/// </summary>
public static class StatusEffects
{
    public static void Apply(Enemy enemy, string kind, double duration, double potency = 0, int stacks = 1)
    {
        bool isBoss = enemy is Beaudis or Dissonance or PathChaseBoss or SinChemesthesisBoss or PhantasiaBoss or PlagueTouchBoss;
        if (kind == "stun")
        {
            enemy.StatusControlResistance = Math.Min(.8, enemy.StatusControlResistance + (isBoss ? .18 : .08));
            duration *= (1 - enemy.StatusControlResistance) * (isBoss ? .45 : 1.0);
        }
        if (!enemy.StatusEffects.TryGetValue(kind, out var effect))
        {
            effect = new StatusEffectState();
            enemy.StatusEffects[kind] = effect;
        }
        effect.Remaining = Math.Max(effect.Remaining, duration);
        effect.Potency = Math.Max(effect.Potency, potency);
        int stackCap = kind switch { "bleed" => 8, "bane" => 30, _ => 3 };
        effect.Stacks = Math.Min(stackCap, effect.Stacks + stacks);
        GameProfile.IncrementQuest("statuses_applied");
    }

    public static void RollPlayerHit(Enemy enemy, Bullet bullet, IEnumerable<ItemDrop?> equipment, double projectileCount,
        Random? rng = null)
    {
        rng ??= Random.Shared;
        var chances = new Dictionary<string, double>(Items.StatusChances(equipment));
        double coefficient = Math.Max(.22, 1 / Math.Sqrt(Math.Max(1, projectileCount)));
        if (bullet.IsCritical)
            chances["bleed"] = chances.GetValueOrDefault("bleed") + .025;
        var tuning = new Dictionary<string, (double Duration, double Potency)>
        {
            ["poison"] = (4.5, .007), ["bleed"] = (3.2, .006), ["slow"] = (2.4, .22),
            ["daze"] = (2.2, .28), ["stun"] = (.65, 1),
        };
        foreach (var (kind, chance) in chances)
        {
            if (tuning.TryGetValue(kind, out var values) && rng.NextDouble() <= Math.Min(.65, chance * coefficient))
                Apply(enemy, kind, values.Duration, values.Potency);
        }
    }

    public static double DamageMultiplier(Enemy enemy, Bullet? bullet = null)
    {
        double multiplier = 1;
        if (enemy.StatusEffects.ContainsKey("poison") && enemy.StatusEffects.ContainsKey("daze"))
            multiplier += .08;
        if (enemy.StatusEffects.ContainsKey("dread"))
            multiplier += .15;
        if (bullet?.IsCritical == true && enemy.StatusEffects.TryGetValue("bleed", out var bleed))
            multiplier += Math.Min(.20, bleed.Stacks * .025);
        // Unlike bleed's crit-only bonus, Bane (Grimsbane's signature stacking
        // curse -- see UniqueEffects.OnPlayerHit's "bane_on_hit") boosts every
        // hit the target takes, +5% per stack, up to the 30-stack cap (+150%).
        if (enemy.StatusEffects.TryGetValue("bane", out var bane))
            multiplier += bane.Stacks * .05;
        return multiplier;
    }

    public static StatusControl Update(Enemy enemy, double seconds)
    {
        enemy.StatusControlResistance = Math.Max(0, enemy.StatusControlResistance - seconds * .035);
        double dotPerSecond = 0, movement = 1, daze = 0;
        bool stunned = false;
        foreach (var (kind, effect) in enemy.StatusEffects.ToArray())
        {
            effect.Remaining -= seconds;
            if (effect.Remaining <= 0)
            {
                enemy.StatusEffects.Remove(kind);
                continue;
            }
            switch (kind)
            {
                case "poison": dotPerSecond += Math.Max(2, enemy.MaxHp * effect.Potency) * effect.Stacks; break;
                case "bleed": dotPerSecond += Math.Max(1, enemy.MaxHp * effect.Potency) * effect.Stacks; break;
                case "slow": movement *= Math.Max(.45, 1 - effect.Potency); break;
                case "dread": movement *= Math.Max(.40, 1 - effect.Potency); break;
                case "daze": daze = Math.Max(daze, effect.Potency); break;
                case "stun": stunned = true; break;
            }
        }
        if (dotPerSecond > 0 && enemy.Hp > 0)
        {
            enemy.StatusDotBuffer += dotPerSecond * seconds;
            int damage = (int)enemy.StatusDotBuffer;
            if (damage > 0)
            {
                enemy.StatusDotBuffer -= damage;
                enemy.TakeDamage(damage, "body", DamageSource.DamageOverTime);
            }
        }
        return new StatusControl(movement, daze, stunned);
    }
}
